### PR TITLE
Support global variable initialization

### PR DESCRIPTION
**Pull request by eecheng87**
In latest version, there already have .data section which is for storing **initialized global variable**. As a result, I don't add .bss section which is for uninitialized variable in this pull request.

### CODE REVIEW

**Review by jserv — CHANGES_REQUESTED** [review]
Append "Close #12" in git commit message(s). Therefore, you can link a pull request to an issue to show that a fix is in progress and to automatically close the issue when the pull request is merged.
See https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue

**Comment by jserv** [comment]
You can remove the statement "2. Global variable initialization is not supported." in `README.md`.

File: tests/global.c [hunk 1]
```diff
@@ -0,0 +1,13 @@
+int a = 44;
```
Comment (jserv) [inline tests/global.c:1]: How about dropping `tests/global.c` and adding the original test in file `tests/driver.sh`?
I am planning to delete some trivial files in directory `tests` such as `def.c`, `fib.c`, `loop.c`, and `pointer.c`, which are already covered in test driver.

Comment (jserv) [reply tests/global.c:1]: However, it still makes sense to generate an executable from `tests/hello.c`.

**Comment by jserv** [comment]
You can prepare another git commit which applies this change to simplify the handing in file `src/globals.c`. The reason why function `global_init` exists is because the global variable initializations were not supported.

File: src/cfront.c [hunk 2]
```diff
@@ -1308,6 +1309,25 @@ int read_body_assignment(char *token, block_t *parent)
...
 }
 
+int read_global_assignment(char *token)
+{
+    var_t *var = find_global_var(token);
+    if (var) {
+        char buffer[10];
+        /* global initialization must be constant */
+        /* TODO: support rvalue as expression e.g. int a = 3 + 2; */
+        if (lex_peek(T_numeric, buffer))
```
Comment (jserv) [inline src/cfront.c:1319]: Can you reuse the code in function `read_body_assignment`? That is, split some logic/operations into a helper function and then invoke the helper for reading the expressions from assignments.

**Comment by jserv** [comment]
> I have changed commit with key word **Close #12** but nothing change. Did it work after PR merge?

If you check the end of #12, the issue is already associated with this pull request. Then, when this pull request is merged, #12 will be closed automatically.

File: src/cfront.c (hunk 2 above)
Comment (eecheng87) [reply src/cfront.c:1319]: Do I need to do that anyway? The reason I don't like to do that is:
* The only duplicate code in `read_body_assignment` is `find_global_var`
* Future update about global initialization will make more different with `read_body_assignment`. If I merge now, I guess `read_body_assignment` will become harder to read
* Merge control flow into `read_body_assignment` for one line re-use makes `read_body_assignment` more complicated

File: tests/global.c (hunk 1 above)
Comment (eecheng87) [reply tests/global.c]: Done! Move global initialization into `tests/driver.sh`

File: README.md [hunk 3]
```diff
@@ -165,19 +165,18 @@ int fib(int n)      fib:                        Reserve stack frame for function
 
 1. Any non-zero value is NOT treated as logical truth by all ops.
    That is, the expression `0 == strcmp(ptr, "hello")` is not equivalent to `!strcmp(ptr, "hello")`.
-2. Global variable initialization is not supported. 
-   Therefore, you can not initialize a global such as `int i = [expr]`.
-3. Dereference is incomplete. Consider `int x = 5; int *ptr = &x;` and it is forbidden to use `*ptr`.
+2. Dereference is incomplete. Consider `int x = 5; int *ptr = &x;` and it is forbidden to use `*ptr`.
    However, it is valid to use `ptr[0]`, which behaves the same of `*ptr`.
-4. The support of varying number of function arguments is incomplete. No `<stdarg.h>` can be used.
+3. The support of varying number of function arguments is incomplete. No `<stdarg.h>` can be used.
```
Comment (jserv) [inline README.md:170]: Don't change the order. Minimize the required changes.

Comment (jserv) [reply README.md:170]: Later, we can create the corresponding issues via GitHub rather than the bullet points.

File: src/cfront.c (hunk 2 above)
Comment (jserv) [reply src/cfront.c:1322]: I agree. Let's stick to the reasonably effective changes.

File: src/cfront.c [hunk 4]
```diff
@@ -1308,6 +1309,30 @@ int read_body_assignment(char *token, block_t *parent)
...
 }
 
+int read_global_assignment(char *token)
+{
+    var_t *var = find_global_var(token);
+    if (var) {
+        char buffer[10];
+        int isneg = 0;
+        /* global initialization must be constant */
+        /* TODO: support rvalue as expression e.g. int a = 3 + 2; */
```
Comment (jserv) [inline src/cfront.c:1319]: This issue should be resolved before this pull request is about to be merged.

**Comment by eecheng87** [comment]
Latest commit support using expression to initialize global variable (also add test bench in `driver.sh`).

`read_global_assignment` using the similar concept with `read_expr` to process expression.
The reason I don't want merge them in one function is that they actually for totally different purpose.

`read_global_assignment` need to know the result before execution time and write into .data section before it.
`read_expr` don't need to get exact result of expression, it only need series instruction and get result during execution time.

File: src/cfront.c [hunk 5]
```diff
@@ -1308,6 +1309,142 @@ int read_body_assignment(char *token, block_t *parent)
...
+    char buffer[10];
+    if (lex_accept(T_minus))
+        isneg = 1;
+    if (lex_peek(T_numeric, buffer))
+        res = read_numeric_constant(buffer);
+    else
+        error("Invalid value after assignment");
+    lex_expect(T_numeric);
+    if (isneg)
+        return -1 * res;
```
Comment (jserv) [inline src/cfront.c:1325]: Rewrite as `(-1) * res`.

File: src/cfront.c [hunk 6]
```diff
@@ -1308,6 +1309,142 @@ int read_body_assignment(char *token, block_t *parent)
...
+    else
+        return res;
+}
+
+int get_expression_imm(opcode_t op, int op1, int op2)
```
Comment (jserv) [inline src/cfront.c:1330]: The verb `get` might not reflect what the function does. It can be `evaluate` or `eval` in short.

File: src/cfront.c [hunk 7]
```diff
@@ -1308,6 +1309,142 @@ int read_body_assignment(char *token, block_t *parent)
...
+        break;
+    case OP_lshift:
+        res = op1 << op2;
+        break;
+    case OP_rshift:
+        res = op1 >> op2;
+        break;
+    default:
+        error("Using operation not support");
+        break;
```
Comment (jserv) [inline src/cfront.c:1355]: No `break` is required. Remove it.

File: README.md [hunk 8]
```diff
@@ -165,8 +165,7 @@ int fib(int n)      fib:                        Reserve stack frame for function
 
 1. Any non-zero value is NOT treated as logical truth by all ops.
    That is, the expression `0 == strcmp(ptr, "hello")` is not equivalent to `!strcmp(ptr, "hello")`.
-2. Global variable initialization is not supported. 
-   Therefore, you can not initialize a global such as `int i = [expr]`.
+2. ELF lacks of .bss and .rodata section
```
Comment (jserv) [inline README.md:168]: Should be "The generated ELF file lacks..."

File: src/cfront.c [hunk 9]
```diff
@@ -1308,6 +1309,140 @@ int read_body_assignment(char *token, block_t *parent)
...
+        res = op1 / op2;
+        break;
+    case OP_lshift:
+        res = op1 << op2;
+        break;
+    case OP_rshift:
+        res = op1 >> op2;
+        break;
+    default:
+        error("Using operation not support");
```
Comment (jserv) [inline src/cfront.c:1353]: Should be "The requested operation is not supported."

**Review by jserv — CHANGES_REQUESTED** [review]
Improve the git commit message. You should address 1) how you made it in summary; 2) Known limitation; 3) New test item.

**Review by jserv — CHANGES_REQUESTED** [review]
Change the message "Add correspond test bench in line 250 of driver.sh" to "Add global initialization item in test driver."

**Review by jserv — CHANGES_REQUESTED** [review]
In git commit messages, always use quotes ("...") instead of backquote. Change "help us evaluate" to "evaluates."
Change "This commit doesn't" to "this patch ..."

**Comment by jserv** [comment]
Thank @eecheng87 for contributing!

**Pull request merged**